src/unix/poll.c

```diff
@@ -84,32 +90,35 @@ define_closure_function(1, 0, void, epoll_free,
     deallocate_vector(e->events);
     if (e->f.close && e->f.close != INVALID_ADDRESS)
         deallocate_closure(e->f.close);
-    unix_cache_free(get_unix_heaps(), epoll, e);
+    deallocate(epoll_heap, e, sizeof(*e));
 }
 
-static epoll epoll_alloc_internal(void)
+static epoll epoll_alloc_internal(int epoll_type)
 {
-    epoll e = unix_cache_alloc(get_unix_heaps(), epoll);
+    epoll e = allocate_zero(epoll_heap, sizeof(struct epoll));
     if (e == INVALID_ADDRESS)
-	return e;
+        return e;
 
+    e->epoll_type = epoll_type;
     list_init(&e->blocked_head);
     init_refcount(&e->refcount, 1, init_closure(&e->free, epoll_free, e));
+    spin_lock_init(&e->blocked_lock);
+    spin_lock_init(&e->fds_lock);
     e->h = heap_general(get_kernel_heaps());
```

Comment:
This should be replaced by the locked heap

Reply:
Fixed!